Update the search-query-syntax examples to show multiple example search terms (#11191)
mlflow/mlflow: docs/source/search-runs.rst

Search Runs
===========

This guide will walk you through how to search your MLflow runs through the MLflow UI and Python API.
This resource will be valuable if you're interested in querying specific runs based on their metrics,
params, tags, dataset information, or run metadata.

Search Runs on MLflow UI
------------------------

The MLflow UI provides a powerful search interface that allows you to filter runs. Below we'll...

> 1. Create example mllfow runs

Reviewer: ```suggestion
1. Create example MLflow runs
```

2. Look at a simple querying example
3. Deep dive into query syntax
4. Provide a variety of example queries

Create Example MLflow Runs
^^^^^^^^^^^^^^^^^^^^^^^^^^

First, let's create some example MLflow runs. This documentation is based on experiments created 
with the below script. If you don't want to interactively explore this on your machine, skip 
this section.

Before running the script, let's simply start the MLflow UI on a local host. 

.. code-block:: bash

  mlflow ui

Let's visit ``http://localhost:5000/`` in our web browser. After doing so, you'll notice that we don't 

> have any experiments or models. Let's resolve this by creating a few MLflow runs. Via the script

Reviewer: ```suggestion
have any experiments or models. Let's resolve this by creating a few MLflow runs via the script
```

below. 

> Note that when you run this script, you'll want to do so from the same directory that your ran

Reviewer: ```suggestion
Note that when you run this script, you'll want to do so from the same directory that you ran
```

          mlflow.log_input(dataset, context=dataset_context[i])

The code above creates 10 MLflow runs with different metrics, params, tags and dataset information. 

> After successful execution, if you return to the MLflow UI in your browser, you should find all 

Reviewer: ```suggestion
After successful execution, if you return to the MLflow UI in your browser, you should find all of 
```

> these runs under the experiment "search-run-guide", as shown by the screenshot:

Reviewer: ```suggestion
these runs under the experiment "search-run-guide", as shown by the following screenshot:
```

.. figure:: _static/images/search-runs/created_mlflow_runs.png
   :alt: testing runs
   :width: 90%
   :align: center

> Now in production systems, it's common to have thousands or even hundreds of thousands of mlflow 

Reviewer: ```suggestion
In real-world production deployments of MLflow, it's common to have thousands or even hundreds of thousands of  
```

runs. In such cases, it's important to be able to filter and search for runs based on specific criteria.

.. _search-runs-syntax:

Search Query Example 
^^^^^^^^^^^^^^^^^^^^

To filter your MLflow runs, you need to write **search queries**, which are SQL-like conditions. 

> To showcase this functionality, let's reopen the MLflow UI, let's explore some example fields that you can use to filter your runs in the MLflow UI.

Reviewer: ```suggestion
To showcase this functionality, let's reopen the MLflow UI and explore some example fields that you can use to filter your runs.
```

.. figure:: _static/images/search-runs/before_search_query.png
   :alt: search query
   :width: 90%

>    :align: center

Reviewer 2: Can we use code example instead a screenshot? Screenshots can be outdated easily.

Reviewer: +1

Second, let's try filtering the runs for our really bad models: ``metrics.loss > 0.8``.

.. figure:: _static/images/search-runs/after_search_query.png
   :alt: search query
   :width: 90%
   :align: center

You'll notice that we now are displaying 2 runs instead of 10. Pretty easy, right?

Now let's go over the search query syntax in more detail.

Search Query Syntax Deep Dive
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

As noted above, MLflow search syntax is similar to SQL with a few notable exceptions.

* The SQL ``OR`` keyword is not supported.

> * For fields that contain special characters or start with numbers, you need to wrap them in double quotes.

Reviewer: Let's make this a callout as well above line 138. This is quite possibly the most frequently overlooked component of this search syntax.

  * Good: ``metrics."cross-entropy-loss" < 0.5``
  * Bad:  ``metrics.cross-entropy-loss < 0.5``
  * Good: ``params."1st_iteration_timestamp" = "2022-01-01"``
  * Bad:  ``params.1st_iteration_timestamp = "2022-01-01"``

> * For the SQL ``IN`` keyword, you must surround the values of your list with single quotes.

Reviewer: For the good / bad examples, can we make it more visually apparent the difference between them?

```rst
.. code-block:: diff

    - Bad: params."learning rate" IN ("0.001", "0.01")
    + Good: params."learning rate" IN ('0.001', '0.01')
```
(might require some adjustment, but is likely to be very obvious at a quick glance which one to pay attention to for a quick reference)

  * Good: ``params."learning rate" IN ('0.001', '0.01')``
  * Bad:  ``params."learning rate" IN ("0.001", "0.01")``

> * For the SQL ``IN`` keyword, you can only search the following fields:

Reviewer: This could use some refinement. Parameters can be searched, right? 

For the datasets.{any_attribute} condition, can an example be provided?

Author: Good catch. Parameters **cannot** be searched with IN

  * ``datasets.{any_attribute}``
  * ``attributes.run_id``

* Non-None conditions for numeric fields are not supported e.g. ``metrics.accuracy != "None"`` will fail.

Other than the that, the syntax should be intuitive to anyone who has used SQL. To assemble
a single search condition, you must assemble an inequality using the following components...

1. **An MLflow field**: a metric, param, tag, dataset or run metadata.
2. **A comparator**: an inequality operator. 

  * For numerics, MLflow supports ``=``, ``!=``, ``>``, ``>=``, ``<``, and ``<=``.

>   * For strings, MLflow supports ``=``, ``!=``, ``LIKE`` (case-sensitive) and ``ILIKE`` (case-insensitive). 

Reviewer: Let's provide some examples of common query terms (particularly LIKE and ILIKE) and some range search conditions (metric_A >= 0.7654, etc.)

>   * For sets, MLflow supports ``IN``.

Reviewer: Let's add an example of searching over a set

3. **A reference value**: a numeric value, string, or set of strings.

Let's look at some examples.

Example Queries
^^^^^^^^^^^^^^^

In this section we will go over how to search by different categories of MLflow fields. For each category we provide
a few sample queries. If you have executed the run creation script we provided, these queries should fetch certain runs
but sometimes require modification for run-specific information, such as ``start_time``.

1 - Searching By Metrics
~~~~~~~~~~~~~~~~~~~~~~~~

Metrics are quantitative measures typically used to evaluate the model's performance during 
or after training. Metrics can include values like accuracy, precision, recall, F1 score, etc., and 
can change over time as the model trains. They are logged manually via ``mlflow.log_metric`` or 
``mlflow.log_metrics`` or automatically via autologging.

To search for runs by filtering on metrics, you must include the ``metrics`` prefix in the left 
side of the inequality. Note that they are **stored as numbers**, so you must use numeric comparators.

.. code-block:: sql

>   metrics.accuracy > 0.72

Reviewer: is metrics."accuracy" > 0.72 valid? Let's show that both are options.

comparators, such as ``=`` and ``!=``.

.. code-block:: sql

    tags."environment" = "notebook"
    tags.environment = "notebook"
    tags.task = "Classification"
    tags.task ILIKE "classif%"

4 - Searching By Dataset Information
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Datasets represent data used in model training or evaluation, including features, targets, 
predictions, and metadata such as the dataset's name, digest (hash) schema, profile, and source. 
They are logged via ``mlflow.log_input`` or automatically via autologging.

To search for runs by filtering on dataset information, you must filter on one of the below fields

1. ``datasets.name``, which is the dataset's name.
2. ``datasets.digest``, which is a unique identifier for the dataset.
3. ``datasets.context``, which represents if the dataset is used for train, evaluation or test.

Note that dataset information is **stored as strings**, so you must use string comparators, such as ``=`` and ``!=``.
Also note that datasets support set comparators, such as ``IN``.

.. code-block:: sql

    datasets.name like "custom"
    datasets.digest IN ('s8ds293b', 'jks834s2')

>     datasets.context = "train"

Reviewer: For line 255 above, can we fix the DML statement to be capitalized? While we do support "like", standard ANSI SQL lint formatting dictates that DML operators are all CAPS

5 - Searching By Run's Metadata
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Run metadata are a variety of user-specified and system-generated attributes that provide additional context about the run.

> To search for runs by filtering on run's metadata, you must include the ``attributes`` prefix in the left

Reviewer: ```suggestion
To search for runs by filtering on the metadata of runs, you must include the ``attributes`` prefix in the left
```

side of the inequality. Note that run metadata can be either a string or a numeric depending on the 

> attribute, so you must use the appropriate comparator. For a complete list of attributes, see

Reviewer: Can we add examples of this?

    AND metrics.loss > 0.3
    AND metrics."f1 score" < 0.5
    AND params.model LIKE "GPT%"
  """

  runs_with_complex_filter = mlflow.search_runs(
      experiment_names=["search-run-guide"],
      filter_string=complex_filter,
  )
  print(runs_with_complex_filter)

The output will be a pandas DataFrame with the runs that match the specified filters, as shown below.

.. code-block:: text

                                run_id  ... tags.mlflow.runName
  0  22db81f070f6413588641c8c343cdd72  ...   orderly-quail-568
  1  c3680e37d0fa44eb9c9fb7828f6b5481  ...    melodic-lynx-301

  [2 rows x 19 columns]

2 - `run_view_type`
~~~~~~~~~~~~~~~~~~~

The ``run_view_type`` parameter exposes additional filtering options, as noted in the 
:py:class:`mlflow.entities.ViewType` enum. For example, if you want to filter only active runs,
which is a dropdown in the UI, simply pass ``run_view_type=ViewType.ACTIVE_ONLY``.

.. code-block:: python

  import mlflow
  from mlflow.entities import ViewType

  active_runs = mlflow.search_runs(
      experiment_names=["search-run-guide"],
      run_view_type=ViewType.ACTIVE_ONLY,
      order_by=["metrics.accuracy DESC"],
  )

2 - Ordering
~~~~~~~~~~~~

> Another useful feature is ordering the results. You specify a list of columns of interest along with

Reviewer: ```suggestion
Another useful feature that is available in the search API is allowing for ordering of the returned search results. You can specify a list of columns of interest along with
```
